### ISSUE TITLE
Out Of Memory Issues

### ISSUE THREAD

**Issue by rlee-uls** [jaerith/ONIX-Data#6]
Hi Jaerith,

Thank you again for help on the other items!  It has been very helpful!

I’m now getting a lot of OutOfMemory errors on big files....  These files are in the 800 MB range. 

The errors happen during the processing in OnixData.
 
Thursday, March 4, 2021 11:18:32 AM: exception processing Single file: System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.String.CtorCharArrayStartLength(Char[] value, Int32 startIndex, Int32 length)
   at System.Xml.XmlTextReaderImpl.get_Value()
   at System.Xml.XmlLoader.LoadNode(Boolean skipOverWhitespace)
   at System.Xml.XmlLoader.LoadDocSequence(XmlDocument parentDoc)
   at System.Xml.XmlLoader.Load(XmlDocument doc, XmlReader reader, Boolean preserveWhitespace)
   at System.Xml.XmlDocument.Load(XmlReader reader)
   at OnixData.OnixParser.get_MessageHeader() in Z:\Visual Studio Projects\ONIX-Data-master\OnixData\OnixParser.cs:line 169
   at OnixData.OnixEnumerator.MoveNext() in Z:\Visual Studio Projects\ONIX-Data-master\OnixData\OnixParser.cs:line 316
 
I was wondering if you had any ideas on what I can do to try and have the full file process.   

Thanks again!

**Comment by jaerith** [comment]
Hi Ordio,

Sure, I'll start looking into it when I have some extra time.

**Comment by WizeLardus** [comment]
I regularly have to process large files from publishers, and ran into the same issue at the start.

I inherited a "proof of concept" project, and saw that the previous DEV had code to do the following: first split the XML file into smaller files. About 500 to 1000 records at a time. He used a trial for a paid for XML parser (obviously did not tell boss so he looked cooler), but I changed it to then process them using this great library.

The idea might assist those running into this issue until such time jaerith has a chance to implement something.

**Comment by rlee-uls** [comment]
Thank you @jaerith, appreciate it a lot!  

Thanks @WizeLardus for that.  I never thought to just split the XML files.
I'll have to look into that and see if can find something quick and easy to split and parse on the fly.

**Comment by jaerith** [comment]
Thanks for the compliments, @ordio and @WizeLardus!

Actually, after finding a better way to handle namespaces, there are now two classes (which are more memory efficient) that you can use for large files: OnixLegacyPlusParser (for ONIX 2.1) and OnixPlusParser (for ONIX 3.0).  I know, they're terrible names for classes; nobody should ever ask me to name their baby.  In any case, they function the same as OnixLegacyParser and OnixParser, and I might include an example of their usage in the README file.

However, there is one caveat to know before using either of them.  For the sake of brevity (since I won't go into the long story of it here), the ONIX-Data project does not use the DTD and performs preprocessing on ONIX data before doing any actual parsing.  With smaller files, it performs the preprocessing in memory, so your actual file is not affected.  However, with larger files, it must perform the preprocessing on the file itself, persisting the changes to the file before it's read and parsed.  These changes are merely real-world substitutions for ONIX encodings (found in the ONIX DTD), which is the same result for the output when parsing with a DTD.  However, in the case of large files, we are actually changing the file itself, not the parsed results.  In any case, this preprocessing can take a few minutes to finish, depending on the machine's specs.  In the case of my machine, preprocessing is normally 6-8 minutes per 400 MB.

So, in summary, if you value the original copy of your large ONIX file, be sure to create a backup copy beforehand.  Please take these classes for a spin and tell me if they work for you.  If there's anything amiss, be sure to let me know!

**Comment by rlee-uls** [comment]
Thanks @jaerith, i'll let you know how it works!

**Comment by rlee-uls** [comment]
Hi @jaerith,

I was able to process a 1GB file (990 MB)! 

So the changes you made has a huge impact on the import.

Thank you very much!!

**Comment by jaerith** [comment]
Great to hear!  I'm glad that it worked for you.
